diffmpm/element.py

```diff
@@ -1,30 +1,539 @@
+import abc
+import itertools
+from typing import Tuple
+
 import jax.numpy as jnp
+from jax import jacobian, jit, lax, vmap
+
+from diffmpm.node import Nodes
+
+
+class _Element(abc.ABC):
+    @abc.abstractmethod
+    def id_to_node_ids(self):
+        ...
+
+    @abc.abstractmethod
+    def id_to_node_loc(self):
+        ...
+
+    @abc.abstractmethod
+    def id_to_node_vel(self):
+        ...
 
-class Bar1D:
+    # @abc.abstractmethod
+    # def tree_flatten(self):
+    #     ...
+
+    # @classmethod
+    # @abc.abstractmethod
+    # def tree_unflatten(self):
+    #     ...
+
+
+class Linear1D(_Element):
     """
-    1D bar element with 2 nodes.
-
-    Attributes:
-
-    id: int
-        Element index.
-    
-    nid1: int
-        Node index 1.
-    
-    nid2: int
-        Node index 2.
-
-    length: float
-        Element length.
-    
-    particles: list
-        List of particles in bar.
+    Container for 1D line elements (and nodes).
+
+    Element ID:            0     1     2     3
+    Mesh:               +-----+-----+-----+-----+
+    Node IDs:           0     1     2     3     4
+
+    + : Nodes
+    +-----+ : An element
     """
 
-    def __init__(self):
-        self.id = None
-        self.node1 = None
-        self.node2 = None
-        self.size = jnp.array([0])
-        self.particles = []
\ No newline at end of file
+    def __init__(self, nelements: int, el_len: float):
+        """Initialize Linear1D.
+
+        Arguments
+        ---------
+        nelements : int
+            Number of elements.
+        el_len : float
+            Length of each element.
+        """
+        self.ids: jnp.ndarray = jnp.arange(nelements)
+        self.el_len: float = el_len
+        self.nodes: Nodes = Nodes(
+            nelements + 1, jnp.arange(nelements + 1).reshape(-1, 1, 1) * el_len
+        )
+
+    def id_to_node_ids(self, id: int):
+        """
+        Node IDs corresponding to element `id`.
+
+        Arguments
+        ---------
+        id : int
+            Element ID.
+
+        Returns
+        -------
+        jax.numpy.ndarray
+            Nodal IDs of the element. Shape of returned
+        array is (2, 1)
+        """
+        return jnp.array([id, id + 1]).reshape(2, 1)
+
+    def id_to_node_loc(self, id: int):
+        """
+        Node locations corresponding to element `id`.
```

Comment:
Is this the physical location of a node? then shouldn't that depend on the element size?

Reply:
That works...(for most cases). So I won't worry about it. In MPM, most often the mesh has the same element size throughout.